### PR TITLE
Support IBMRestServlet in JAX-RS 3.0

### LINKED ISSUE #12875: Enable IBMRestServlet when using RESTEasy feature (part of JAX-RS 3.0 prototyping)
Allow users to continue specifying the IBMRestServlet in their web.xml in JAX-RS 3.0 (based on RESTEasy).

Some possible caveats:
* Looks like RESTEasy can only handle one Application subclass per app/web-module
* Only one "REST" servlet (IBMRestServlet or the servlets provided by RESTEasy)
* The REST servlet must only be mapped to a single path mapping in the web.xml
* The Application subclass must define the providers and resources - if there is an Application subclass, things don't work if that subclass does not override the getClasses and getSingletons methods.

### PR DESCRIPTION
**Pull request by andymc12**
Fixes #12875

Implement IBMRestServlet in JAX-RS 3.0:
- adds new ServletContainerInitializer that auto-adds some web.xml config
- adds new IBMRestServlet that just extends HttpServlet30Dispatcher
- new messages when multiple REST servlets or mappings configured in same module
- exposes RESTEasy's META-INF/services files to TCCL
 - Move JAX-RS 2.1 META-INF/services files from API jar to impl jar to avoid duplicate entries while the 3.0 prototyping continues to use the 2.1 APIs

### CODE REVIEW

**Comment by andymc12** [comment]
#build

File: dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+#-------------------------------------------------------------------------------
+# # {0} description of each insert field
+# MSG_DESCRIPTIVE_NAME_CWMGQ1xxxI=CWMGQ1xxxI: This is a message with inserts {0}
+# MSG_DESCRIPTIVE_NAME_CWMGQ1xxxI.explanation=Explanation text for the message
+# MSG_DESCRIPTIVE_NAME_CWMGQ1xxxI.useraction=User action text for the message
+#
+#-------------------------------------------------------------------------------
+
+MULTIPLE_REST_SERVLETS_CWWKW1300W=CWWKW1300W: Multiple REST servlets are defined for the web module, {0}. Only zero or one REST servlet is allowed per web module.
+MULTIPLE_REST_SERVLETS_CWWKW1300W.explanation=Only one REST servlet is allowed to be defined per web module. A REST servlet is any servlet that uses one of the following class names: com.ibm.websphere.jaxrs.server.IBMRestServlet, org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher, or org.jboss.resteasy.plugins.server.servlet.HttpServlet30Dispatcher. This message indicates that more than one is specified either in the web.xml or configured dynamically.
```
Comment (ManasiGandhi) [inline dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops:56]: Here it says,
"This message indicates that more than one is specified either in the web.xml or configured dynamically. "

Can you elaborate on what "more than one" is? Is it the servlet?

File: dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+# MSG_DESCRIPTIVE_NAME_CWMGQ1xxxI.explanation=Explanation text for the message
+# MSG_DESCRIPTIVE_NAME_CWMGQ1xxxI.useraction=User action text for the message
+#
+#-------------------------------------------------------------------------------
+
+MULTIPLE_REST_SERVLETS_CWWKW1300W=CWWKW1300W: Multiple REST servlets are defined for the web module, {0}. Only zero or one REST servlet is allowed per web module.
+MULTIPLE_REST_SERVLETS_CWWKW1300W.explanation=Only one REST servlet is allowed to be defined per web module. A REST servlet is any servlet that uses one of the following class names: com.ibm.websphere.jaxrs.server.IBMRestServlet, org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher, or org.jboss.resteasy.plugins.server.servlet.HttpServlet30Dispatcher. This message indicates that more than one is specified either in the web.xml or configured dynamically. 
+MULTIPLE_REST_SERVLETS_CWWKW1300W.useraction=Review the web.xml for these servlets and remove all or all but one.
+
+MULTIPLE_REST_SERVLET_MAPPINGS_CWWKW1301W=CWWKW1301W: Multiple REST servlet mappings are defined for the web module, {0}. A REST servlet may only be associated with a single path mapping.
```
Comment (ManasiGandhi) [inline dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops:59]: Change,
"Multiple REST servlet mappings are defined for the web module, {0}. A REST servlet may only be associated with a single path mapping."
to,
"Multiple REST servlet mappings are defined for the web module, {0}. A REST servlet can be associated only with a single path mapping."

File: dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+MULTIPLE_REST_SERVLET_MAPPINGS_CWWKW1301W.explanation=A REST servlet is only allowed to be configure with zero or one path mapping. A REST servlet is any servlet that uses one of the following class names: com.ibm.websphere.jaxrs.server.IBMRestServlet, org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher, or org.jboss.resteasy.plugins.server.servlet.HttpServlet30Dispatcher. This message indicates that more than one mapping is associated with a REST servlet in the web.xml.
```
Comment (ManasiGandhi) [inline dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops:60]: Change,
"A REST servlet is only allowed to be configure with zero or one path mapping."
to,
"A REST servlet is only allowed to be configured with zero or one path mapping. "

File: dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+MULTIPLE_REST_SERVLETS_CWWKW1300W.useraction=Review the web.xml for these servlets and remove all or all but one.
+
+MULTIPLE_REST_SERVLET_MAPPINGS_CWWKW1301W=CWWKW1301W: Multiple REST servlet mappings are defined for the web module, {0}. A REST servlet may only be associated with a single path mapping.
+MULTIPLE_REST_SERVLET_MAPPINGS_CWWKW1301W.explanation=A REST servlet is only allowed to be configure with zero or one path mapping. A REST servlet is any servlet that uses one of the following class names: com.ibm.websphere.jaxrs.server.IBMRestServlet, org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher, or org.jboss.resteasy.plugins.server.servlet.HttpServlet30Dispatcher. This message indicates that more than one mapping is associated with a REST servlet in the web.xml. 
+MULTIPLE_REST_SERVLET_MAPPINGS_CWWKW1301W.useraction=Review the web.xml for the servlet-mappings associated with the REST servlet and remove all or all but one.
```
Comment (ManasiGandhi) [inline dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops:61]: Change,
"Review the web.xml for the servlet-mappings associated with the REST servlet and remove all or all but one."
to,
"Review the web.xml for the servlet-mappings that are associated with the REST servlet and remove all or all but one."

File: dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops (hunk 1 above)
Comment (andymc12) [reply dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops:56]: Hi @ManasiGandhi - that is correct.  It means more than one servlet is specified.

File: dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops (hunk 2 above)
Comment (andymc12) [reply dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops:59]: Will do.

File: dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops (hunk 3 above)
Comment (andymc12) [reply dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops:60]: Good catch - will do.

File: dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops (hunk 4 above)
Comment (andymc12) [reply dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops:61]: Will do.

**Comment by andymc12** [comment]
Last build timed out accessing RTC - new #build

**Comment by andymc12** [comment]
@ManasiGandhi all suggested NLS props file changes are in. Would you take another look for approval?  Thanks!

File: dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops (hunk 1 above)
Comment (ManasiGandhi) [inline dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops:56]: Change,
"This message indicates that more than one is specified either in the web.xml or configured dynamically. "
to,
"This message indicates that more than one servlet is specified either in the web.xml or configured dynamically. "

Comment (andymc12) [reply dev/io.openliberty.org.jboss.resteasy.common/resource/io/openliberty/org/jboss/resteasy/common/nls/JAXRSServer.nlsprops]: Done - thanks again!

**Comment by andymc12** [comment]
message update and rebased from master - new #build

**Comment by ManasiGandhi** [comment]
Changes look god. Thanks. Approved.
 
----- Original message -----From: Andy McCright <notifications@github.com>To: OpenLiberty/open-liberty <open-liberty@noreply.github.com>Cc: Manasi Gandhi <manasi.gandhi@ibm.com>, Mention <mention@noreply.github.com>Subject: [EXTERNAL] Re: [OpenLiberty/open-liberty] Support IBMRestServlet in JAX-RS 3.0 (#12876)Date: Tue, Jul 7, 2020 9:54 AM 
 
@ManasiGandhi all suggested NLS props file changes are in. Would you take another look for approval? Thanks!

**Comment by andymc12** [comment]
All failures occur in many other builds (without this change) and do not seem related to this change.
#run-libby-bot

**Pull request merged**